File: core/java/src/main/java/hara/lang/protocol/IPromise.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** An asynchronous value with composable settlement handlers. */
@HaraProtocolBinding(
    namespace = "std.protocol.ipromise",
    name = "IPromise",
    parents = {"IDeref", "IDerefTimeout"})
public interface IPromise
    extends IDeref<Object>, IDerefTimeout<Object> {
  @HaraMethod(value = "state", arity = 1)
  Object state();

  @HaraMethod(value = "value", arity = 1)
  Object value();

  @HaraMethod(value = "then", arity = 2)
  Object then(Object function);

  @HaraMethod(value = "catch", arity = 2)
  Object catchError(Object function);

  @HaraMethod(value = "finally", arity = 2)
  Object finallyDo(Object function);

  @HaraMethod(value = "cancel", arity = 1)
  Object cancel();
}

File: core/rust/src/lang/protocol/ipromise.rs
use super::{IDeref, IDerefTimeout};
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.ipromise",
    name = "IPromise",
    parents = ["IDeref", "IDerefTimeout"]
)]
pub trait IPromise<V, F>: IDeref<Output = V> + IDerefTimeout<V> + Sized {
    type State;
    type Error;

    #[hara_method(value = "state", arity = 1)]
    fn state(&self) -> Self::State;
    #[hara_method(value = "value", arity = 1)]
    fn value(&self) -> Result<V, Self::Error>;
    #[hara_method(value = "then", arity = 2)]
    fn then(&self, function: F) -> Self;
    #[hara_method(value = "catch", arity = 2)]
    fn catch(&self, function: F) -> Self;
    #[hara_method(value = "finally", arity = 2)]
    fn r#finally(&self, function: F) -> Self;
    #[hara_method(value = "cancel", arity = 1)]
    fn cancel(&self) -> bool;
}
